pkg/shardddl/optimism/lock.go

```diff
@@ -263,11 +278,23 @@ func (l *Lock) TrySync(info Info, tts []TargetTable) (newDDLs []string, err erro
 			// for these two cases, we should execute the DDLs to the downstream to update the schema.
 			log.L().Info("joined table info changed", zap.String("lock", l.ID), zap.Int("cmp", cmp), zap.Stringer("from", oldJoined), zap.Stringer("to", newJoined),
 				zap.String("source", callerSource), zap.String("schema", callerSchema), zap.String("table", callerTable), zap.Strings("ddls", ddls))
-			// check for add column with a larger field len
 			if cmp < 0 {
-				_, err = AddDifferentFieldLenColumns(l.ID, ddls[idx], oldJoined, newJoined)
-				if err != nil {
-					return ddls, err
+				// check for add column with a larger field len
+				if col, err2 := AddDifferentFieldLenColumns(l.ID, ddls[idx], oldJoined, newJoined); err2 != nil {
+					return ddls, cols, err2
+				} else if len(col) > 0 && l.IsDroppedColumn(info.Source, info.UpSchema, info.UpTable, col) {
+					return ddls, cols, terror.ErrShardDDLOptimismTrySyncFail.Generate(
+						l.ID, fmt.Sprintf("add column %s that wasn't fully dropped in downstream. ddl: %s", col, ddls[idx]))
+				}
+			} else {
+				if col, err2 := GetColumnName(l.ID, ddls[idx], ast.AlterTableDropColumn); err2 != nil {
+					return ddls, cols, err2
+				} else if len(col) > 0 {
+					err = l.AddDroppedColumn(info, col)
```

Comment:
from the aspect of "cols is a cache of etcd", i think we should return when error or don't append to cols